src/crypto/index.js

```diff
@@ -271,61 +285,74 @@ function _maybeUploadOneTimeKeys(crypto) {
 
     crypto._lastOneTimeKeyCheck = now;
 
-    function uploadLoop(numberToGenerate) {
-        if (numberToGenerate <= 0) {
+    // We need to keep a pool of one time public keys on the server so that
+    // other devices can start conversations with us. But we can only store
+    // a finite number of private keys in the olm Account object.
+    // To complicate things further then can be a delay between a device
+    // claiming a public one time key from the server and it sending us a
+    // message. We need to keep the corresponding private key locally until
+    // we receive the message.
+    // But that message might never arrive leaving us stuck with duff
+    // private keys clogging up our local storage.
+    // So we need some kind of enginering compromise to balance all of
+    // these factors.
+
+    // Check how many keys we can store in the Account object.
+    const maxOneTimeKeys = crypto._olmDevice.maxNumberOfOneTimeKeys();
+    // Try to keep at most half that number on the server. This leaves the
+    // rest of the slots free to hold keys that have been claimed from the
+    // server but we haven't recevied a message for.
+    // If we run out of slots when generating new keys then olm will
+    // discard the oldest private keys first. This will eventually clean
+    // out stale private keys that won't receive a message.
+    const keyLimit = Math.floor(maxOneTimeKeys / 2);
+
+    function uploadLoop(keyCount) {
+        if (keyLimit <= keyCount) {
             // If we don't need to generate any more keys then we are done.
             return;
         }
 
-        const keysThisLoop = Math.min(numberToGenerate, maxKeysPerCycle);
+        const keysThisLoop = Math.min(keyLimit - keyCount, maxKeysPerCycle);
 
         // Ask olm to generate new one time keys, then upload them to synapse.
         crypto._olmDevice.generateOneTimeKeys(keysThisLoop);
-        return _uploadOneTimeKeys(crypto).then(() => {
-            return uploadLoop(numberToGenerate - keysThisLoop);
+        return _uploadOneTimeKeys(crypto).then((res) => {
+            if (res.one_time_key_counts && res.one_time_key_counts.signed_curve25519) {
+                // if the response contains a more up2date value use this
+                // for the next loop
+                return uploadLoop(res.one_time_key_counts.signed_curve25519);
+            } else {
+                return uploadLoop(keyCount + keysThisLoop);
```

Comment:
if the response fields aren't present, doesn't that suggest something has gone horribly wrong? it might be better to give up with an error, than to carry on regardless